airbyte-lib/airbyte_lib/caches/base.py

```diff
@@ -0,0 +1,717 @@
+            max_batch_id = max(batches_to_finalize.keys())
+
+            # Make sure the target schema and target table exist.
+            self._ensure_schema_exists()
+            final_table_name = self._ensure_final_table_exists(
+                stream_name,
+                create_if_missing=True,
+            )
+            self._ensure_compatible_table_schema(
+                stream_name=stream_name,
+                table_name=final_table_name,
+                raise_on_error=True,
+            )
+
+            try:
+                temp_table_name = self._write_files_to_new_table(
+                    files,
+                    stream_name,
+                    max_batch_id,
+                )
+                self._write_temp_table_to_final_table(
+                    stream_name,
+                    temp_table_name,
+                    final_table_name,
+                )
+            finally:
+                self._drop_temp_table(temp_table_name, if_exists=True)
+
+            # Return the batch handles as measure of work completed.
+            return batches_to_finalize
+
+    def _execute_sql(self, sql: str | TextClause | Executable) -> CursorResult:
+        """Execute the given SQL statement."""
+        if isinstance(sql, str):
+            sql = text(sql)
+        if isinstance(sql, TextClause):
+            sql = sql.execution_options(
+                autocommit=True,
+            )
+
+        with self.get_sql_connection() as conn:
+            try:
+                result = conn.execute(sql)
+            except (
+                sqlalchemy.exc.ProgrammingError,
+                sqlalchemy.exc.SQLAlchemyError,
+            ) as ex:
+                msg = f"Error when executing SQL:\n{sql}\n{type(ex).__name__}{ex!s}"
+                raise SQLRuntimeError(msg) from None  # from ex
+
+        return result
+
+    def _drop_temp_table(
+        self,
+        table_name: str,
+        if_exists: bool = True,
+    ) -> None:
+        """Drop the given table."""
+        exists_str = "IF EXISTS" if if_exists else ""
+        self._execute_sql(f"DROP TABLE {exists_str} {self._fully_qualified(table_name)}")
+
+    def _write_files_to_new_table(
+        self,
+        files: list[Path],
+        stream_name: str,
+        batch_id: str,
+    ) -> str:
+        """Write a file(s) to a new table.
+
+        This is a generic implementation, which can be overridden by subclasses
+        to improve performance.
+        """
+        temp_table_name = self._create_table_for_loading(stream_name, batch_id)
+        for file_path in files:
+            with pa.parquet.ParquetFile(file_path) as pf:
+                record_batch = pf.read()
+                dataframe = record_batch.to_pandas()
+
+                # TODO: Add back metadata columns (this breaks tests)
+                # dataframe["_airbyte_loaded_at"] = str(pd.Timestamp.now())
+
+                dataframe.to_sql(
```

Comment:
I tested with a source that has date columns and the final resulting table has a varchar again.
I think it's caused by this logic - the pandas dataframe has the date column as a string and via the `if_exists="replace"`, it won't honor the real table type and fall back to whatever types the dataframe provides which won't have gone through our sql type converter logic.

So in summary, it will create the table with the right types via `_create_table_for_loading`, then throw away these types again during the pandas loading step.

When switching `if_exists` to `append`, it works as expected (as the table is empty before this `to_sql` call)

Reply:
Thanks for the clarification. The pandas "to_sql" was a bit of a hack, written before we had any catalog or type information.

Reply:
Found and fixed an error where pandas defaulted to "replace" instead of "append". Also was able to plug in our detected datatypes into the Pandas "to_sql" implementation - although in practice this should be a no-op because now the previous table creation method will not be replaced. I considered just letting Pandas write the table, but Pandas doesn't seem to support Primary Keys, which we'll probably want eventually.

Reply:
I think in a future iteration, we can add more advanced tests to ensure all data types and columns match exactly, but for that to be worthwhile we also need more complex test datasets, including datasets with diverse types and PK configurations. I think this is sufficient for now, since we're just trying to meet the functional spec of writing the records and being able to read them back.

Also, this PR is huge and I don't want to add more scope.